File: Data Structure/Lab 15/InfixtoPrefix.c
// Write a program to convert infix notation to prefix notation using stack.

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

#define N 100

char S[N];
int top = -1; 

int IPF(char ch) {
    if (ch == '+' || ch == '-') {
        return 1;
    }
    if (ch == '*' || ch == '/') {
        return 3;
    }
    if (ch == '^') {
        return 6;
    }
    if (ch == ')') {
        return 0;
    }
    if (ch == '(') {
        return 9;
    }
    return -1;
}

int SPF(char ch) {
    if (ch == '+' || ch == '-') {
        return 2;
    }
    if (ch == '*' || ch == '/') {
        return 4;
    }
    if (ch == '^') {
        return 5;
    }
    if (ch == ')') {
        return 0;
    }
    if (ch == '(') {
        return 0;
    }
    return 0;
}

void PUSH(char x) {
    if (top < N - 1) {
        S[++top] = x;
    } else {
        printf("\nStack Overflow!\n");
    }
}

char POP() {
    if (top >= 0) {
        return S[top--];
    }
    return '\0';
}

void reverseString(char *str) {
    int len = strlen(str);
    for (int i = 0; i < len / 2; i++) {
        char temp = str[i];
        str[i] = str[len - 1 - i];
        str[len - 1 - i] = temp;
    }
}

void Conversion(const char *input) {
    char INFIX[N];
    char POLISH[N] = "\0";
    int p_index = 0;
    
    strcpy(INFIX, input);

    reverseString(INFIX);

    int len = strlen(INFIX);
    for (int i = 0; i < len; i++) {
        if (INFIX[i] == '(') {
            INFIX[i] = ')';
        } else if (INFIX[i] == ')') {
            INFIX[i] = '(';
        }
    }
    
    top = -1; 
    int index = 0;
    
    while (INFIX[index] != '\0') {
        char Next = INFIX[index];
        
        if (isalnum(Next)) {
            POLISH[p_index++] = Next;
            POLISH[p_index] = '\0';
        } else if (Next == '(') {
            PUSH(Next);
        } else if (Next == ')') {
            while (top >= 0 && S[top] != '(') {
                POLISH[p_index++] = POP();
                POLISH[p_index] = '\0';
            }
            POP(); 
        } else { 
            while (top >= 0 && SPF(S[top]) > IPF(Next)) { 
                POLISH[p_index++] = POP();
                POLISH[p_index] = '\0';
            }
            PUSH(Next);
        }
        index++;
    }
    
    while (top >= 0) {
        POLISH[p_index++] = POP();
        POLISH[p_index] = '\0';
    }
  
    reverseString(POLISH);

    printf("Result :\n");
    printf("Infix expression  : %s\n", input);
    printf("Prefix expression : %s\n", POLISH);
    fflush(stdout); 
}

int main() {
    char exp[N];
    printf("Enter infix exp : ");
    fflush(stdout);
    
    if (scanf("%99s", exp) == 1) {
        Conversion(exp);
    }
    return 0;
}

File: Data Structure(Java)/Lab 13/InfixtoPrefix.java
// Write a program to convert infix notation to prefix notation using stack.

import java.util.Scanner;

public class InfixtoPrefix{

    static int N = 100;
    static char[] S = new char[N];
    static int top = -1; 
    
    static int IPF(char ch){
        if(ch == '+' || ch == '-'){
            return 1;
        }
        if(ch == '*' || ch == '/'){
            return 3;
        }
        if(ch == '^'){
            return 6;
        }
        if(Character.isLetterOrDigit(ch)){
            return 7;
        }
        if(ch =='('){
            return 9;
        }
        if(ch == ')'){
            return 0;
        }
        return -1;
    }
    
    static int SPF(char ch){
        if (ch == '+' || ch == '-'){
            return 2;
        }
        if (ch == '*' || ch == '/'){
            return 4;
        }
        if (ch == '^'){
            return 5;
        }
        if (Character.isLetterOrDigit(ch)){
            return 8; 
        }
        if (ch == '('){
            return 0;
        }
        return 0;
    }
    
    static int RANK(char ch) {
        if (ch == '+' || ch == '-' || ch == '*' || ch == '/' || ch == '^') {
            return -1;
        }
        if (Character.isLetterOrDigit(ch)) {
            return 1;
        }
        return 0;
    }
    
    static void PUSH(char ch){
        if(top < N - 1){
            S[++top] = ch;
        }
        else{
            System.out.println("Stack overflow!!");
            return;
        }
    }
    
    static char POP(){
        if(top == -1){
            return '\0';
        }
        else{
            return S[top--];
        }
    }
    
    static void Conversion(char[] input){
        int len = input.length;
        char[] INFIX = new char[len + 2];

        for(int i = 0; i <len ; i++){
            char raw = input[len - 1 - i];
            INFIX[i] = (raw == '(') ? ')' : (raw == ')') ? '(' : raw;
        }

        int p_index = 0;
        INFIX[len] = ')';
        INFIX[len + 1] = '\0';
        top = -1; 

        PUSH('(');
        char[] POLISH = new char[N];
        
        int rank = 0;
        int index = 0;
        char NEXT = INFIX[index++];
        
        while (NEXT != '\0') {
            if(top < 0){ 
                System.out.println("Stack underflow!!");
                return;
            }
            
            if(Character.isLetterOrDigit(NEXT)){
                POLISH[p_index++] = NEXT;
                rank = rank + RANK(NEXT);
            }
            else{

                while(top >= 0 && SPF(S[top]) >= IPF(NEXT)){
                    char temp = POP();
                    
                    if (temp != '(' && temp != ')') {
                        POLISH[p_index++] = temp;
                        rank = rank + RANK(temp);
                    }
                    if(rank < 0){ 
                        System.out.println("Rank can't be less than 0. Current rank :"+rank);
                        return;
                    }
                }
                
                if(top >= 0 && SPF(S[top]) != IPF(NEXT)){
                    PUSH(NEXT);
                }
                else{
                    POP();
                }
            }
            
            NEXT = INFIX[index++]; 
        }
        
        if(top != -1 || rank != 1){ 
            System.out.println("ERROR : Final evaluation is Invalid : Top = " + top +" Rank = "+ rank);
        }
        else{
            System.out.println("Infix expression :"+new String(input));
            System.out.print("Prefix expression :");

            for(int i = p_index - 1; i >= 0; i--) {
                System.out.print(POLISH[i]);
            }

            System.out.println();
        }
    }
    
    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        
        System.out.print("Enter input exp :");
        String inputstr = sc.next();

        int len = inputstr.length();
        char[] input = new char[len];

        for(int i = 0 ; i < len ; i++){
            input[i] = inputstr.charAt(i);
        }

        Conversion(input);
        
        sc.close();
    }
}
